lib/charms/openstack_libs/v0/gnocchi_requires.py

```diff
@@ -29,4 +31,91 @@
 # to 0 if you are raising the major API version
 LIBPATCH = 1
 
-# TODO: add your code here! Happy coding!
+
+class GnocchiConnectedEvent(EventBase):
+    """Gnocchi connected Event."""
+    pass
+
+
+class GnocchiReadyEvent(EventBase):
+    """Gnocchi ready for use Event."""
+    pass
+
+
+class GnocchiGoneAwayEvent(EventBase):
+    """Gnocchi relation has gone-away Event."""
+    pass
+
+
+class GnocchiServerEvents(ObjectEvents):
+    """Events class for `on`."""
+
+    connected = EventSource(GnocchiConnectedEvent)
+    ready = EventSource(GnocchiReadyEvent)
+    goneaway = EventSource(GnocchiGoneAwayEvent)
+
+
+class GnocchiRequires(Object):
+    """Requires side interface for gnocchi interface type."""
+
+    on = GnocchiServerEvents()
+
+    def __init__(
+        self, charm, relation_name: str
+    ):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+
+        self.framework.observe(
+            self.charm.on[relation_name].relation_joined,
+            self._on_gnocchi_relation_joined,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_changed,
+            self._on_gnocchi_relation_changed,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_departed,
+            self._on_gnocchi_relation_changed,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_broken,
+            self._on_gnocchi_relation_broken,
+        )
+
+    def _on_gnocchi_relation_joined(self, event):
+        """Gnocchi relation joined."""
+        logging.debug("Gnocchi on_joined")
+        self.on.connected.emit()
+
+    def _on_gnocchi_relation_changed(self, event):
+        """Gnocchi relation changed."""
+        logging.debug("Gnocchi on_changed")
+        self.on.ready.emit()
```

Comment:
I think you only want to emit the ready event when the `gnocchi_url` is available on the relation? With it as it is right now, this will emit an event whenever the relation data changes. In practice this is when the `gnocchi_url`, but there might be some wonky scenarios where some other value is set on the relation (say for testing or debug purposes) and we wouldn't want the event indicating that gnocchi is ready if the necessary information is not there.

Reply:
Added a try and except to check if `gnocchi_url` is available, thanks !